modules/core/src/main/java/org/apache/synapse/message/processor/impl/forwarder/ForwardingService.java

```diff
@@ -609,13 +533,190 @@ public void dispatch(MessageContext messageContext) {
+				isSuccessful = false;
+				onForwardFailure();
+				//invoke fault sequence of MP
+				sendThroughFaultSeq(outCtx);
+			}
+
+			if (!isResponseValidationNotRequired && outCtx != null && validateResponse(outCtx)) {
+				isSuccessful = false;
+				onForwardFailure();
+				//invoke fault sequence of MP
+				sendThroughFaultSeq(outCtx);
+			}
+		}
+	}
+
+	/**
+	 * Steps after successfully forwarding the message to backend
+	 * @param endpoint
+	 */
+	private void onForwardSuccess(Endpoint endpoint) {
+		messageConsumer.ack();
+		attemptCount = 0;
+		isSuccessful = true;
+
+		if (log.isDebugEnabled()) {
+			log.debug("Successfully forwarded the message to endpoint [" +
+					endpoint.getName() + "]" + " with message processor [" +
+					messageProcessor.getName() + "].");
+		}
+	}
+
+	/**
+	 * Steps after failure to forward the message to the BE (service level or connection level)
+	 */
+	private void onForwardFailure() {
+		log.error("BlockingMessageSender of message processor [" +
+				this.messageProcessor.getName() +
+				"] failed to forward message to endpoint " + messageProcessor.getTargetEndpoint());
+	}
+
+	/**
+	 * Validate if response can be considered as a successful Back-end invocation
+	 *
+	 * @param responseMessage MessageContext of response
+	 * @return true if it is a successful invocation
+	 */
+	private boolean validateResponse(MessageContext responseMessage) {
+    	boolean isSuccessful;
+		String responseSc = ((Axis2MessageContext) responseMessage).
+                getAxis2MessageContext().getProperty(SynapseConstants.HTTP_SC).toString();
+		// Some events where response code is null (i.e. sender socket timeout
+		// when there is no response from endpoint)
+		int sc = 0;
+		try {
+            sc = Integer.parseInt(responseSc.trim());
+            isSuccessful = getHTTPStatusCodeFamily(sc).equals(
+                    HTTPStatusCodeFamily.SUCCESSFUL) ||
+                    isNonRetryErrorCode(responseSc);
+        } catch (NumberFormatException nfe) {
+            isSuccessful = false;
+        }
+        if(!isSuccessful) {
+			String statusCode = " ";
+			if(sc != 0 ) {
+				statusCode = Integer.toString(sc);
+			}
+			log.info("Message processor [" + this.messageProcessor.getName() + "] received a response with HTTP_SC: "
+					+ statusCode + " from backend " + targetEndpoint + ". Message forwarding failed.");
+		}
+        return isSuccessful;
+	}
+
+	/**
+	 * Update Axis2MessageContext before dispatching message
+	 *
+	 * @param messageToDispatch message to be forwarded
+	 */
+	private void updateAxis2MessageContext(MessageContext messageToDispatch) {
+		OMElement firstChild = null; //
+		org.apache.axis2.context.MessageContext origAxis2Ctx =
+                ((Axis2MessageContext) messageToDispatch).getAxis2MessageContext();
+		if (JsonUtil.hasAJsonPayload(origAxis2Ctx)) {
+            firstChild = origAxis2Ctx.getEnvelope().getBody().getFirstElement();
```

Comment:
is this a formatting issue?

Reply:
looks some are tabs some are 4 spaces

Reply:
Will Fix. Yes, that was the issue with auto format code with IDE.